Common/Statistics/TradeBuilder.cs

```diff
@@ -0,0 +1,486 @@
+        private readonly FillGroupingMethod _groupingMethod;
+        private readonly FillMatchingMethod _matchingMethod;
+        private bool _liveMode;
+
+        private const int LiveModeMaxTradeCount = 10000;
+        private const int LiveModeMaxTradeAgeMonths = 12;
+
+        /// <summary>
+        /// Initializes a new instance of the TradeBuilder class
+        /// </summary>
+        public TradeBuilder(FillGroupingMethod groupingMethod, FillMatchingMethod matchingMethod)
+        {
+            _groupingMethod = groupingMethod;
+            _matchingMethod = matchingMethod;
+        }
+
+        /// <summary>
+        /// Sets the live mode flag
+        /// </summary>
+        /// <param name="live">The live mode flag</param>
+        public void SetLiveMode(bool live)
+        {
+            _liveMode = live;
+        }
+
+        /// <summary>
+        /// The list of closed trades
+        /// </summary>
+        public List<Trade> ClosedTrades
+        {
+            get { return _closedTrades; }
+        }
+
+        /// <summary>
+        /// Returns true if there is an open position for the symbol
+        /// </summary>
+        /// <param name="symbol">The symbol</param>
+        /// <returns>true if there is an open position for the symbol</returns>
+        public bool HasOpenPosition(string symbol)
+        {
+            Position position;
+            if (!_positions.TryGetValue(symbol, out position)) return false;
+
+            if (_groupingMethod == FillGroupingMethod.FillToFill)
+                return position.PendingTrades.Count > 0;
+
+            return position.PendingFills.Count > 0;
+        }
+
+        /// <summary>
+        /// Sets the current market price for the symbol
+        /// </summary>
+        /// <param name="symbol"></param>
+        /// <param name="price"></param>
+        public void SetMarketPrice(string symbol, decimal price)
+        {
+            Position position;
+            if (!_positions.TryGetValue(symbol, out position)) return;
+
+            if (price > position.MaxPrice)
+                position.MaxPrice = price;
+            else if (price < position.MinPrice)
+                position.MinPrice = price;
+        }
+
+        /// <summary>
+        /// Processes a new fill, eventually creating new trades
+        /// </summary>
+        /// <param name="fill">The new fill order event</param>
+        /// <param name="conversionRate">The current market conversion rate into the account currency</param>
+        public void ProcessFill(OrderEvent fill, decimal conversionRate)
+        {
+            // If we have multiple fills per order, we assign the order fee only to its first fill
+            // to avoid counting the same order fee multiple times.
+            var orderFee = 0m;
+            if (!_ordersWithFeesAssigned.Contains(fill.OrderId))
+            {
+                orderFee = fill.OrderFee;
+                _ordersWithFeesAssigned.Add(fill.OrderId);
```

Comment:
I think it's safe to do in live or backtest, but I may be mistaken. My thought is that this hash is really only used to prevent duplicating order fees, i.e. multiple fills that each have an order fee component. I think it's a safe assumption that a single order won't have two fills separated by 20k other orders. So I think we can always used the `FixedSizeHashQueue<T>` class, but either way is fine for me. My main concern here is certainly live mode stability.